### PR TITLE
Fix path joining and quoting in authn/authz

### PR DESCRIPTION
**Pull request by savonarola**
Fixes https://emqx.atlassian.net/browse/EMQX-9600, https://github.com/emqx/emqx/issues/10411

## PR Checklist

- [x] Added tests for the changes
- [x] Changed lines covered in coverage report
- [x] Change log has been added to `changes/{ce,ee}/(feat|perf|fix)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
~~- [ ] If there should be document changes, a PR to emqx-docs.git is sent, or a jira ticket is created to follow up~~
- [x] Schema changes are backward compatible

### CODE REVIEW

File: changes/ce/fix-10420.zh.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+在身份验证和授权模块中为 HTTP 请求编写 URL 时修复 HTTP 路径处理。
+* 避免不必要的 URL 规范化，因为我们不能假设外部服务器平等对待原始 URL 和规范化 URL。 这导致了像 [#10411](https://github.com/emqx/emqx/issues/10411) 这样的错误。
+* 修复路径段可能被 HTTP 编码两次的问题。
```
Comment (zmstone) [inline changes/ce/fix-10420.zh.md:3]: no need for this anymore.

File: apps/emqx_connector/src/emqx_connector_http.erl [hunk 2]
```diff
@@ -553,15 +553,41 @@ formalize_request(Method, BasePath, {Path, Headers, _Body}) when
...
+    lists:reverse(Path1) ++ [$/ | Path2];
+%% "abc" + "/cde"
+do_join_paths(Path1, [$/ | Path2]) ->
+    lists:reverse(Path1) ++ [$/ | Path2];
+%% "abc" + "cde"
+do_join_paths(Path1, Path2) ->
+    lists:reverse(Path1) ++ [$/ | Path2].
+
+to_list(List) when is_list(List) -> List;
+to_list(Bin) when is_binary(Bin) -> binary_to_list(Bin).
```
Comment (zmstone) [inline apps/emqx_connector/src/emqx_connector_http.erl:584]: we have to handle unicode here.

File: apps/emqx_connector/src/emqx_connector_http.erl [hunk 3]
```diff
@@ -553,15 +553,41 @@ formalize_request(Method, BasePath, {Path, Headers, _Body}) when
...
+
+to_bin(Bin) when is_binary(Bin) ->
     Bin;
-bin(Str) when is_list(Str) ->
+to_bin(Str) when is_list(Str) ->
     list_to_binary(Str);
```
Comment (zmstone) [inline apps/emqx_connector/src/emqx_connector_http.erl:589]: unicode list.

File: apps/emqx_authn/src/simple_authn/emqx_authn_http.erl [hunk 4]
```diff
@@ -285,9 +285,9 @@ parse_url(Url) ->
                     BaseUrl = iolist_to_binary([Scheme, "//", HostPort]),
                     case string:split(Remaining, "?", leading) of
                         [Path, QueryString] ->
-                            {BaseUrl, Path, QueryString};
```
Comment (thalesmg) [inline apps/emqx_authn/src/simple_authn/emqx_authn_http.erl:288]: Not that it should be done in this PR, but I feel this `parse_url/1` fn could be replaced by `uri_string:parse` :thinking: 

(same for `apps/emqx_authz/src/emqx_authz_http.erl`)

File: apps/emqx_connector/src/emqx_connector_http.erl (hunk 2 above)
Comment (savonarola) [reply apps/emqx_connector/src/emqx_connector_http.erl:584]: Valid URLs cannot contain wide characters anyway; a request with Unicode will either fail in the library or be rejected by the server.
However, checking URL validity on each request seems to be an overkill 🤔

File: apps/emqx_authn/src/simple_authn/emqx_authn_http.erl (hunk 4 above)
Comment (savonarola) [reply apps/emqx_authn/src/simple_authn/emqx_authn_http.erl:288]: Unfortunately not 😅 This is an old conceptual issue — that users provide _templates_ for urls `https://my.site/user/${username}?password=${password}` which are not valid urls and are treated as invalid by URL parsers.

Comment (savonarola) [reply apps/emqx_authn/src/simple_authn/emqx_authn_http.erl:288]: Seems that this problem was the reason for appearing of this loose parser.

File: apps/emqx_connector/src/emqx_connector_http.erl (hunk 3 above)
Comment (savonarola) [reply apps/emqx_connector/src/emqx_connector_http.erl:589]: This was intended to be a no-op for joining HTTP paths.
`BasePath` is parsed by `emqx_http_lib` and is already a list.
The second argument comes from `on_query` arguments, we may add a guard there to force connector users to take care of converting path to a valid string themselves.

File: changes/ce/fix-10420.zh.md (hunk 1 above)
Comment (savonarola) [reply changes/ce/fix-10420.zh.md]: ✅

**Review by zmstone — APPROVED** [review]

File: apps/emqx_connector/src/emqx_connector_http.erl [hunk 5]
```diff
@@ -642,4 +668,21 @@ redact_test_() ->
...
+        ?_assertEqual("abc/cde", join_paths("abc/", "/cde")),
+
+        ?_assertEqual("/", join_paths("", "")),
+        ?_assertEqual("/cde", join_paths("", "cde")),
+        ?_assertEqual("/cde", join_paths("", "/cde")),
+        ?_assertEqual("/cde", join_paths("/", "cde")),
+        ?_assertEqual("/cde", join_paths("/", "/cde")),
+
+        ?_assertEqual("//cde/", join_paths("/", "//cde/")),
+        ?_assertEqual("abc///cde/", join_paths("abc//", "//cde/"))
```
Comment (HJianBo) [inline apps/emqx_connector/src/emqx_connector_http.erl:685]: nit: why not `abc////cde/`? @savonarola

**Pull request merged**